index.js

```diff
@@ -125,6 +136,460 @@
+      }
+    }
+    return upcarrots + '\n' + highlightNumbers;
+  };
+
+  //  sigGetName :: Node -> String
+  var sigGetName = function(sig) {
+    return sig.children[SIG_IDX_NAME].children[0].text;
+  };
+
+  //  findConstraint :: (TypeClass, String) -> Node -> Nullable Node
+  var findConstraint = function(typeClass, typeVarName) {
+    return function(sig) {
+      return traverseFind(function(node, parents) {
+        if (isVar(node) && node.text === typeVarName) {
+          var parent = getParent(parents);
+          if (parent.text === getConstraintText(typeClass)) {
+            return parent;
+          }
+        }
+      }, sig.children[SIG_IDX_CONSTRAINTS]);
+    };
+  };
+
+  //  findTypeVar :: (String, Node) -> Nullable Node
+  var findTypeVar = function(typeVarName, startNode) {
+    return traverseFind(function(node, parents) {
+      if (isVar(node) && node.text === typeVarName) {
+        return node;
+      }
+    }, startNode);
+  };
+
+  //  findArg :: (Integer, Node) -> Node
+  var findArg = function(index, sig) {
+    var args = nonDelimiters(sig.children[SIG_IDX_ARGS].children);
+    return args[index];
+  };
+
+  //  findArgTypeVar :: (Integer, String) -> Node -> Nullable Node
+  var findArgTypeVar = function(index, typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, findArg(index, sig));
+    };
+  };
+
+  //  findVal :: Integer -> Node -> Node
+  var findVal = function(index) {
+    return function(sig) {
+      return isNaN(index) ? sig.children[SIG_IDX_RET] :
+                            findArg(index, sig);
+    };
+  };
+
+  //  findRetTypeVar :: String -> Node -> Node
+  var findRetTypeVar = function(typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, sig.children[SIG_IDX_RET]);
+    };
+  };
+
+  //  pipe :: (Any -> [(Any -> Any)]) -> Any
+  var pipe = function(data, fns) {
+    var r = fns[0](data);
+    for (var idx = 1; idx < fns.length; idx += 1) {
+      r = fns[idx](r);
+    }
+    return r;
+  };
+
+  //  head :: [a] -> a?
+  var head = function(xs) {
+    return xs[0];
+  };
+
+  //  tail :: [a] -> [a]
+  var tail = function(xs) {
+    return xs.slice(1);
+  };
+
+  //  last :: [a] -> a?
+  var last = function(xs) {
+    return xs[xs.length - 1];
+  };
+
+  //  toPairs :: StrMap a -> [Pair String a]
+  var toPairs = function(obj) {
+    var pairs = [];
+    for (var k in obj) {
+      pairs.push([k, obj[k]]);
+    }
+    return pairs;
+  };
+
+  //  isFirstChild :: (Node, Node) -> Boolean
+  var isFirstChild = function(node, parent) {
+    return head(nonDelimiters(parent.children)) === node;
+  };
+
+  //  isRootFirstChild :: (Node, [Node]) -> Boolean
+  var isRootFirstChild = function(node, parents) {
+    return (
+      isRootChild(parents) &&
+      isFirstChild(node, head(parents))
+    );
+  };
+
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = last;
+
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = isEmpty;
```

Comment:
> Maybe a special case for "collection" (Array, StrMap) types?

We could list each value on a separate line. If the first argument is a list of integers from 1 to 100, say, and the second is `['foo', 'bar', 'baz']`, we could state:

    concat :: Array a -> Array a -> Array a
                    ^          ^
                    1          2

    1)  1 :: Number, Integer
        2 :: Number, Integer
        ...
        99 :: Number, Integer
        100 :: Number, Integer

    2)  "foo" :: String
        "bar" :: String
        "baz" :: String

    Since there is no type of which all the above values are members, the type-variable constraint has been violated.

I'm not sure whether the `...` should be literal or whether we should display all the value–types pairs regardless of how many there are.